README.md

```diff
@@ -414,6 +413,85 @@ Volume requests embedded in Pod spec are provisioned as ephemeral volumes. The v
 ephemeral volume usage.
+
+### Capacity-aware pod scheduling
+
+PMEM-CSI implements the CSI `GetCapacity` call, but Kubernetes
+currently doesn't call that and schedules pods onto nodes without
+being aware of available storage capacity on the nodes. The effect is
+that pods using volumes with late binding may get tentatively assigned
+to a node and then get stuck because that decision is not reconsidered
+when the volume cannot be created there ([a
+bug](https://github.com/kubernetes/kubernetes/issues/72031)). Even if
+that decision is reconsidered, the same node may get selected again
+because Kubernetes does not get informed about the insufficient
+storage. Pods with ephemeral inline volumes always get stuck because
+the decision to use the node [is final](https://github.com/kubernetes-sigs/descheduler/issues/62).
+
+Work is [under
+way](https://github.com/kubernetes/enhancements/pull/1353) to enhance
+scheduling in Kubernetes. In the meantime, PMEM-CSI provides two components
+that help with pod scheduling:
+
+#### Scheduler extender
+
+When a pod requests the special [extended
+resource](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#extended-resources)
+called `pmem-csi.intel.com/scheduler`, the Kubernetes scheduler calls
+a [scheduler
+extender](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/scheduler_extender.md)
+provided by PMEM-CSI with a list of nodes that a pod might run
+on. This extender is implemented in the master controller and thus can
+connect to the controller on each of these nodes to check for
+capacity. PMEM-CSI then filters out all nodes which currently do not
+have enough storage left for the volumes that still need to be
+created. This considers inline ephemeral volumes and all unbound
+volumes, regardless whether they use late binding or immediate
+binding.
+
+This special scheduling can be requested manually by adding this snippet
+to one container in the pod spec:
+```
+containers:
+- name: some-container
+  ...
+  resources:
+    limits:
+      pmem-csi.intel.com/scheduler: "1"
+    requests:
+      pmem-csi.intel.com/scheduler: "1"
+```
+
+This scheduler extender is optional and not necessarily installed in
+all clusters that have PMEM-CSI. Don't add this extended resource
+unless the scheduler extender is installed, otherwise the pod won't
+start!
+
+#### Pod admission webhook
+
+Having to add `pmem-csi.intel.com/scheduler` manually is not
+user-friendly. To simplify this, PMEM-CSI provides a [mutating
+admission
+webhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/)
+which intercepts the creation of all pods. If that pod uses volumes
+provided by PMEM-CSI, the webhook transparently adds the extended
+resource request.
+
+The webhook can only do that if the persistent volume claim (PVC) and
+its storage class have been created already. This is normally not
```

Comment:
Doe this webook also mutates the Pods which are using neither late-binding or ephemeral volumes but just claims with immediate binding?

Reply:
It mutates all pods using PMEM-CSI. It intentionally checks as little as possible, because between the situation might theoretically still change between pod creation and pod scheduling.

The scheduler extender then looks at volumes and storage classes more thoroughly (https://github.com/intel/pmem-csi/blob/d1c38b6a1c4e9b96f7dd2bc8905675974f9df022/pkg/scheduler/scheduler.go#L223-L227).

Reply:
> It mutates all pods using PMEM-CSI.

I  wouldn't feel like we should touch the Pods using normal persistent volumes, those should go via the normal scheduling process. 

> It intentionally checks as little as possible,

If I am not wrong there are not many extra checks needed except checking the binding mode of the storage class linked to the volume claim, as anyway need the storage class at this point to check if it's provisioned by us.

>  because the situation might theoretically still change between pod creation and pod scheduling.

I won't expect if there would be any change in binding mode between those phases.

Reply:
Okay, I made change in a new commit.